# Exploring eBay Car Sales Data
This project explores a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

# This is for later
print(autos['registration_month'].value_counts(dropna=False))
unique_zip = np.sort(autos['postal_code'].unique())
print(unique_zip[:20])

In [4]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='latin1')

In [5]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [6]:
print(autos.info())
#print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

Apparently some columns have null values (vehicleType, gearbox, model, fuelType, notRepairedDamage). Columns consist of objects or integers.

## Cleaning Column Names

In [7]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [8]:
new_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'n_pictures', 'postal_code',
       'last_seen']
autos.columns = new_columns
print(autos.head())

          date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158 

What I've done here is shortened some of the columns names for readability/typability and changed the titles from camelcase to Python's preferred snakecase to be more universally applicable.
To do this quickly, I leveraged the ```df.columns``` attribute to show the current column names, and then copied and pasted that array so I could update the names.

## Initial Exploration and Cleaning

In [9]:
print(autos.info())
print(autos.describe(include='all'))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  object
 5   ab_test             50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   power_ps            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer            50000 non-null  object
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

We can see that 'price' and 'odemeter' can be converted to numeric types once cleaned.

The columns'seller', 'offer_type', 'ab_test', and 'gearbox' appear to only have a few unique values. Since the columns mostly repeats themselves, they likely don't contain much useful data and are candidates for dropping.

The column 'registration_month' needs further investigation since the minimum month is 0 and the most frequent is NaN.
The 'postal_code' column needs a similar lookover since the minimum is only 4 digits rather than the typical 5.

### Handling numeric data stored as text
We'll start with cleaning the numeric data that's stored as text so we can convert it to numeric type.

In [10]:
# Price
# remove dollar sign and commas and convert to float
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '')
autos['price'] = autos['price'].astype(float)

# rename column
autos.rename(columns={'price' : 'price_usd'}, inplace=True)
print(autos.describe(include='all'))

               date_crawled         name  seller offer_type     price_usd  \
count                 50000        50000   50000      50000  5.000000e+04   
unique                48213        38754       2          2           NaN   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot           NaN   
freq                      3           78   49999      49999           NaN   
mean                    NaN          NaN     NaN        NaN  9.840044e+03   
std                     NaN          NaN     NaN        NaN  4.811044e+05   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.100000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  1.000000e+08   

       ab_test vehicle_type  registration_year  gearbox      power_ps  mode

In [11]:
# Odemeter
# remove 'km' and commas and convert to int
autos['odometer'] = autos['odometer'].str.replace('km', '').str.replace(',', '')
autos['odometer'] = autos['odometer'].astype(int)

# rename column
autos.rename(columns={'odometer' : 'odometer_km'}, inplace=True)

### Deeper exploration into cleaned columns
Let's check the price and odometer columns to (1) verify they are formatted as expected and (2) check for any outliers.

In [12]:
print(autos['price_usd'].unique().shape)
print(autos['price_usd'].describe())
print(autos['price_usd'].value_counts(dropna=False).head(10))
print(autos['price_usd'].value_counts(dropna=False).tail(10))

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64
price_usd
0.0       1421
500.0      781
1500.0     734
2500.0     643
1000.0     639
1200.0     639
600.0      531
800.0      498
3500.0     498
2000.0     460
Name: count, dtype: int64
price_usd
79933.0    1
4895.0     1
33777.0    1
39800.0    1
83000.0    1
16845.0    1
15470.0    1
18890.0    1
5198.0     1
1960.0     1
Name: count, dtype: int64


There are many listing with a price of 0, and it possible to have a free car listed, so we'll leave these. However, the maximum price listed is $100,000,000, which is far above the mean and even far above the third quartile. To get rid of these sorts of outliers, we'll keep only the data for which the price is within two standard deviations of the mean.

In [13]:
mean_price_plus_2std = autos['price_usd'].mean() + 2*autos['price_usd'].std()
autos = autos[autos['price_usd'].between(0, mean_price_plus_2std)]

Let's look at the remaining price data.


In [14]:
print(autos['price_usd'].unique().shape)
print(autos['price_usd'].describe())
print(autos['price_usd'].value_counts(dropna=False).head(10))
print(autos['price_usd'].value_counts(dropna=False).tail(10))

(2347,)
count     49986.000000
mean       5721.525167
std        8983.617820
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      350000.000000
Name: price_usd, dtype: float64
price_usd
0.0       1421
500.0      781
1500.0     734
2500.0     643
1200.0     639
1000.0     639
600.0      531
800.0      498
3500.0     498
2000.0     460
Name: count, dtype: int64
price_usd
1870.0      1
11687.0     1
145000.0    1
130000.0    1
385.0       1
22200.0     1
1497.0      1
16995.0     1
1330.0      1
4567.0      1
Name: count, dtype: int64


Now let's do a similar analysis for the odometer column.

In [15]:
print(autos['odometer_km'].unique().shape)
print(autos['odometer_km'].describe())
print(autos['odometer_km'].value_counts(dropna=False))

(13,)
count     49986.000000
mean     125736.506222
std       40038.133399
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
odometer_km
150000    32416
125000     5169
100000     2168
90000      1757
80000      1436
70000      1230
60000      1164
50000      1025
5000        966
40000       818
30000       789
20000       784
10000       264
Name: count, dtype: int64


All the values in the odometer column seem reasonable and within the same distribution. The maximum value is within two standard deviations (actually within one) of the mean. I don't see any real outliers.

### Date columns

In [16]:
print(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False))
print(autos['date_crawled'].str[:10].sort_values())

date_crawled
2016-04-03    0.038691
2016-03-20    0.037831
2016-03-21    0.037490
2016-03-12    0.036770
2016-03-14    0.036630
2016-04-04    0.036490
2016-03-07    0.035970
2016-04-02    0.035410
2016-03-19    0.034910
2016-03-28    0.034850
2016-03-29    0.034150
2016-03-15    0.033990
2016-04-01    0.033809
2016-03-30    0.033629
2016-03-08    0.033269
2016-03-09    0.033209
2016-03-22    0.032909
2016-03-26    0.032489
2016-03-11    0.032489
2016-03-23    0.032389
2016-03-10    0.032129
2016-03-31    0.031909
2016-03-25    0.031749
2016-03-17    0.031509
2016-03-27    0.031049
2016-03-16    0.029508
2016-03-24    0.029108
2016-03-05    0.025387
2016-03-13    0.015564
2016-03-06    0.013944
2016-04-05    0.013104
2016-03-18    0.013064
2016-04-06    0.003181
2016-04-07    0.001420
Name: proportion, dtype: float64
30296    2016-03-05
43158    2016-03-05
2720     2016-03-05
32747    2016-03-05
32740    2016-03-05
            ...    
49569    2016-04-07
47885    2016-04-07
38743    201

In [17]:
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False))
print(autos['ad_created'].str[:10].sort_values())

ad_created
2016-04-03    0.038931
2016-03-20    0.037871
2016-03-21    0.037691
2016-04-04    0.036850
2016-03-12    0.036610
                ...   
2016-01-16    0.000020
2016-02-07    0.000020
2015-12-30    0.000020
2016-01-07    0.000020
2016-02-16    0.000020
Name: proportion, Length: 76, dtype: float64
22781    2015-06-11
20649    2015-08-10
34883    2015-09-09
2243     2015-11-10
27986    2015-12-05
            ...    
23409    2016-04-07
361      2016-04-07
39445    2016-04-07
13976    2016-04-07
44797    2016-04-07
Name: ad_created, Length: 49986, dtype: object


In [18]:
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False))
print(autos['last_seen'].str[:10].sort_values())

last_seen
2016-04-06    0.220982
2016-04-07    0.130957
2016-04-05    0.124275
2016-03-17    0.027928
2016-04-03    0.025367
2016-04-02    0.024887
2016-03-30    0.024847
2016-04-04    0.024627
2016-03-31    0.023827
2016-03-12    0.023807
2016-04-01    0.023106
2016-03-29    0.022326
2016-03-22    0.021586
2016-03-28    0.020846
2016-03-21    0.020726
2016-03-20    0.020706
2016-03-24    0.019565
2016-03-25    0.019205
2016-03-23    0.018585
2016-03-26    0.016965
2016-03-16    0.016445
2016-03-27    0.016024
2016-03-15    0.015884
2016-03-19    0.015744
2016-03-14    0.012804
2016-03-11    0.012524
2016-03-10    0.010763
2016-03-09    0.009843
2016-03-13    0.008983
2016-03-08    0.007582
2016-03-18    0.007422
2016-03-07    0.005362
2016-03-06    0.004421
2016-03-05    0.001080
Name: proportion, dtype: float64
33684    2016-03-05
17497    2016-03-05
34139    2016-03-05
19738    2016-03-05
22493    2016-03-05
            ...    
28105    2016-04-07
14064    2016-04-07
14069    2016-0

In [19]:
print(autos['registration_year'].describe())

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


Clearly there are some issues with some of the entries for registration year since the minimum year is 1000 and the maximum is 9999, neither of which are within the current modern era of vehicles.

Since a car can't have a registration date after the listing was seen, the oldest year of registration must be 2016. The mass production of cars was expanded by Henry Ford in around 1913, so we can't have any registration years prior to that. Let's remove any data points that do not fall within this range.

In [20]:
autos = autos[autos['registration_year'].between(1913, 2016)]
print(autos['registration_year'].describe())
print(autos['registration_year'].value_counts(normalize=True))

count    48007.000000
mean      2002.823401
std          7.195534
min       1927.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64
registration_year
2000    0.069865
2005    0.062803
1999    0.062449
2004    0.057013
2003    0.056804
          ...   
1943    0.000021
1939    0.000021
1952    0.000021
1938    0.000021
1953    0.000021
Name: proportion, Length: 77, dtype: float64


## Exploring Price by Brand

In [21]:
print(autos['brand'].value_counts())
print(autos['brand'].value_counts(normalize=True))

brand
volkswagen        10185
bmw                5283
opel               5192
mercedes_benz      4579
audi               4149
ford               3350
renault            2273
peugeot            1418
fiat               1242
seat                873
skoda               770
mazda               727
nissan              725
smart               668
citroen             668
toyota              599
sonstige_autos      518
hyundai             473
volvo               444
mini                415
mitsubishi          391
honda               377
kia                 341
alfa_romeo          318
porsche             293
suzuki              284
chevrolet           274
chrysler            176
daihatsu            123
dacia               123
jeep                108
subaru              105
land_rover           98
saab                 77
jaguar               76
trabant              74
daewoo               72
rover                65
lancia               52
lada                 29
Name: count, dtype: int64
brand
vo

Let's analyze the brands that make up more than 5% of car listings.

In [28]:
brand_mean_prices = {}
selected_brands = autos['brand'].value_counts().index[:7]

for b in selected_brands:
    brand_prices = autos.loc[autos['brand'] == b, 'price_usd']
    brand_mean_prices[b] = brand_prices.mean().item()
    
print(sorted(brand_mean_prices.items(), key=lambda item: item[1], reverse=True))

[('audi', 9093.65003615329), ('mercedes_benz', 8485.239571958942), ('bmw', 8102.536248343744), ('volkswagen', 5231.081983308787), ('ford', 3652.095223880597), ('opel', 2877.7224576271187), ('renault', 2396.2067751869777)]


In [29]:
brand_mean_odo = {}

for b in selected_brands:
    brand_odo = autos.loc[autos['brand'] == b, 'odometer_km']
    brand_mean_odo[b] = brand_odo.mean().item()
    
print(sorted(brand_mean_odo.items(), key=lambda item: item[1], reverse=True))

[('bmw', 132431.38368351315), ('mercedes_benz', 130856.0821139987), ('audi', 129287.78018799711), ('opel', 129224.7688751926), ('volkswagen', 128724.10407461954), ('renault', 128238.01143862736), ('ford', 124068.65671641791)]


In [33]:
bmp_series = pd.Series(brand_mean_prices)
bmo_series = pd.Series(brand_mean_odo)

df = pd.DataFrame(bmp_series, columns=['mean_price'])
df['mean_mileage'] = bmo_series

print(df.sort_values('mean_price', ascending=False))
print(df.sort_values('mean_mileage'))

                mean_price   mean_mileage
audi           9093.650036  129287.780188
mercedes_benz  8485.239572  130856.082114
bmw            8102.536248  132431.383684
volkswagen     5231.081983  128724.104075
ford           3652.095224  124068.656716
opel           2877.722458  129224.768875
renault        2396.206775  128238.011439
                mean_price   mean_mileage
ford           3652.095224  124068.656716
renault        2396.206775  128238.011439
volkswagen     5231.081983  128724.104075
opel           2877.722458  129224.768875
audi           9093.650036  129287.780188
mercedes_benz  8485.239572  130856.082114
bmw            8102.536248  132431.383684
